### PR TITLE
gadgets: Update top_file to use map iterators instead of toppers

### PR DESCRIPTION
**Pull request by blanquicet**
# Update top_file gadget to use map iterators instead of toppers

Use map iterators datasource introduced in https://github.com/inspektor-gadget/inspektor-gadget/pull/3182 to run top_file gadget.

Ref https://github.com/inspektor-gadget/inspektor-gadget/issues/1924

## Testing done

Tests where implemented using new testing framework.

## TODO (In follow-up PRs)

- [ ] It's still missing to merge outcomes from nodes (combiner operator): https://github.com/inspektor-gadget/inspektor-gadget/pull/3068
- [ ] `--max-rows` support is still missing (new operator or probably in the combiner)
- [ ] Once agree on this implementation, implement other toppers based on this.

### CODE REVIEW

File: pkg/gadgets/run/types/metadata.go [hunk 1]
```diff
@@ -36,6 +36,9 @@ const (
 	// Prefix used to mark snapshotters structs
 	snapshottersPrefix = "gadget_snapshotter_"
 
+	// Prefix used to mark mapIters info
+	mapIterInfoPrefix = "gadget_mapiter_"
+
```
Comment (blanquicet) [inline pkg/gadgets/run/types/metadata.go:41]: I'm not sure we want to add this commit?

File: gadgets/top_file/test/top_file_test.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+	Writes     uint64 `json:"writes"`
+	ReadBytes  uint64 `json:"rbytes"`
+	WriteBytes uint64 `json:"wbytes"`
+
+	FileType  string `json:"t"`
+	FileName  string `json:"file"`
+	FileInode uint64 `json:"inode"`
+
+	// Device might be 0, so we can't use NormalizedInt and NormalizeInt()
+	// TODO: Support checking for the presence of a field, even if it's 0
```
Comment (burak-ok) [inline gadgets/top_file/test/top_file_test.go:51]: Had the same issue in https://github.com/inspektor-gadget/inspektor-gadget/pull/3115#discussion_r1667073483

File: gadgets/top_file/test/top_file_test.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+			func(t *testing.T, output string) {
+				expectedEntry := &topFileEntry{
+					CommonData: utils.BuildCommonData(containerName, commonDataOpts...),
+
+					// Workload writes "foo" (3 bytes) with "echo" (so "sh") command
+					// into a regular file ('R') named "bar".
+					Comm:       "sh",
+					FileType:   "R",
+					FileName:   "bar",
+					Writes:     1,
```
Comment (burak-ok) [inline gadgets/top_file/test/top_file_test.go:119]: This is highly likely - but is this guaranteed?

File: gadgets/top_file/gadget.yaml [hunk 4]
```diff
@@ -3,56 +3,76 @@ description: Periodically report read/write activity by file
...
+          columns.width: 8
+          columns.alignment: right
+      wbytes:
+        annotations:
+          description: Bytes written by file
+          columns.width: 8
+          columns.alignment: right
+      t:
+        annotations:
+          description: Type of file, 'R' for regular files, 'S' for sockets and 'O'
```
Comment (burak-ok) [inline gadgets/top_file/gadget.yaml:46]: Could he replace the enum values in the eBPF code? So instead of `R` it says `REGULAR`?

**Review by burak-ok — COMMENTED** [review]
Looks good so far, but want to first research the map iterator PR that was merged

File: gadgets/top_file/test/top_file_test.go [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+	igrunner "github.com/inspektor-gadget/inspektor-gadget/pkg/testing/ig"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/testing/match"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/testing/utils"
+	eventtypes "github.com/inspektor-gadget/inspektor-gadget/pkg/types"
+)
+
+type topFileEntry struct {
+	eventtypes.CommonData
+
+	MountNsID uint64 `json:"mntns_id"`
```
Comment (mauriciovasquezbernal) [inline gadgets/top_file/test/top_file_test.go:35]: nit: Use `MntNsID` and done in other gadgets.

File: gadgets/top_file/test/top_file_test.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+	Pid  uint32 `json:"pid"`
+	Tid  uint32 `json:"tid"`
+	Comm string `json:"comm"`
```
Comment (mauriciovasquezbernal) [inline gadgets/top_file/test/top_file_test.go:39]: Can we update the gadget to provide uid / gid as well? See #3169

File: pkg/gadgets/run/types/metadata.go (hunk 1 above)
Comment (mauriciovasquezbernal) [reply pkg/gadgets/run/types/metadata.go:41]: You already implemented it, it works fine, so let's add it.

**Review by mauriciovasquezbernal — APPROVED** [review]
It's looking very good. I only have minor comments.

I tried with  `--map-fetch-count` and it didn't seem to work at the beginning, but then I noticed it has the same issue of #3033. I also noticed #3234, but it's completely unrelated to this PR. 

I think my only blocker is the lack of uid/gid fields. But it's also true it's not related to this PR. So up to you if you want to fix or just merge it.

Thanks!

File: gadgets/top_file/gadget.yaml (hunk 4 above)
Comment (mauriciovasquezbernal) [reply gadgets/top_file/gadget.yaml:46]: I think it's a good point, I'm not sure R, S, O are that easy to understand.

Comment (blanquicet) [reply gadgets/top_file/gadget.yaml:46]: I'll open an issue for this.

**Pull request merged**